proto-grpc.md

```diff
@@ -0,0 +1,277 @@
+convert `.proto` files into a wide variety of languages and libraries.
+
+This page is expected to continue as a 'live' document; as the ecosystem evolves and new rulesets
+are released, this document can be edited at
+[SIG-rules-authors](https://github.com/bazel-contrib/SIG-rules-authors/tree/gh-pages) to continue to
+reflect today's state.
+
+
+## The Basics
+
+At its most basic level, the task of working with Protobuf and gRPC `.proto` files is much the same
+as outside of Bazel: we execute the Protobuf Compiler `protoc` with the necessary arguments,
+environment and plugins to get it to produce a desired set of files or libraries for any specific
+language. For those unfamiliar with protoc, it is a small compiler executable that parses sets of
+`.proto` formatted files and executes plugins to write a set of output source code files. These
+files implement the interfaces for working with the messages and services in the language supported
+by the plugin.
+
+However, given the prevalence of Protobuf and gRPC use, users are not expected to reinvent the rules
+needed for calling protoc themselves. Instead, the common base dependency ruleset
+[bazelbuild/rules_proto](https://github.com/bazelbuild/rules_proto) provides the `proto_library`
+rule which can be used to establish libraries of `.proto` files and map their dependencies. For
+example, suppose we have the following dependency tree of `.proto` files:
+
+![Tree of proto files](./assets/proto-tree.png)
+
+For this set of files, we could define the following collection of `proto_library` targets. Note
+that some rulesets introduced later, such as stackb/rules_proto and rules_go, provide tooling
+through Gazelle for auto-generating these `proto_library` from the `.proto` files.
+
+```starlark
+load("@rules_proto//proto:defs.bzl", "proto_library")
+
+proto_library(
+  name = "proto_a",
+  srcs = ["A.proto"],
+)
+
+proto_library(
+  name = "proto_b",
+  srcs = ["B.proto"],
+)
+
+proto_library(
+  name = "proto_c",
+  srcs = ["C.proto"],
+  deps = [":proto_a", ":proto_b"], # Provides the depended files when calling protoc
+)
+```
+
+Using `bazel query`, we can view the graph that Bazel has assembled from the targets, showing it
+correctly reflects the dependency of `proto_c` on `proto_a` and `proto_b` (you must have
+DOT/GraphViz installed):
+
+`bazel query --output graph --nograph:factored 'kind(proto_library, //...)'  | dot -Tpng > graph.png`
+
+![Generated graph of proto_library targets](./assets/proto-bazel-query.png)
+
+
+Note that we could have grouped all of these `.proto` files under a single `proto_library` in this
+particular case. In real codebases these files may be under separate directories and managed by
+different groups or companies. The convention for splitting or grouping libraries varies by target
+language, but for most it is a matter of local preference. However, for some languages, files with
```

Comment:
I think the general recommendation is that there should be one `proto_library` target for every `.proto` file (can't find the link with that recommendation now, though).

The only exception that I know of is Go, where you need to group them by `go_package` option, not by proto package. I would mention this exception, but also mention that single-file `proto_library` targets should be preferred wherever possible to get better incrementality and more parallel builds.

Reply:
Ok, I'll restructure this recommendation. Is Go actually an exception to this rule? I know that everything under the same `go_package` must be linked under the same final `go_proto_library` (vaguely similar to ODR in C++), but we can still have multiple `proto_libraries` as input to that target? If anything, I was under the impression Go and Gazelle actually pushed for target per proto?

Reply:
Yes, Go is an exception. You actually need an explicit `go_proto_library` graph next to the `proto_library` graph because `go_proto_library` doesn't use an aspect to propagate along the graph, and also duplicate `deps`. The `proto_library` to generate code for is passed as `proto` attribute, which accepts exactly one target. Last time I checked, Gazelle simply put all `.proto` files in a directory into one target (which produces a completely broken build in, e.g., `googleapis` ...)

Reply:
You can choose the mode.  By default is will put everything into the same target, but there is a `package` and now `file` (which works with googleapis).

https://github.com/bazelbuild/bazel-gazelle/blob/master/language/proto/config.go#L105-L121